File: exercises/cash.c
#include <cs50.h>
#include <stdio.h>

int calculate_fraction(int cents, int fraction) { return cents / fraction; }

int calculate_remaining_cents(int cents, int fraction) {
  return cents % fraction;
}

int main(void) {
  const int DIME = 10;
  const int NICKEL = 5;
  const int PENNY = 1;
  const int QUARTER = 25;

  int cents = 0;
  do {
    cents = get_int("Change owed: ");
  } while (cents <= 0);

  int remaining_cents = cents;
  int result = 0;

  result += calculate_fraction(remaining_cents, QUARTER);
  remaining_cents = calculate_remaining_cents(remaining_cents, QUARTER);

  result += calculate_fraction(remaining_cents, DIME);
  remaining_cents = calculate_remaining_cents(remaining_cents, DIME);

  result += calculate_fraction(remaining_cents, NICKEL);
  remaining_cents = calculate_remaining_cents(remaining_cents, NICKEL);

  result += calculate_fraction(remaining_cents, PENNY);
  remaining_cents = calculate_remaining_cents(remaining_cents, PENNY);

  printf("%i\n", result);
}

File: exercises/cash.py
import math


def main():
    change = get_change()
    print(get_money_count(change))


def get_change() -> float:
    while True:
        try:
            change = float(input("Change: "))
            if change > 0:
                return change
        except ValueError:
            pass


def calc_change_and_fractions(change: float, currency: float) -> tuple[float, int]:
    fractions = math.floor(change / currency)
    change = round(change - fractions * currency, 2)

    return change, fractions


def get_money_count(change: float) -> int:
    currencies = [
        0.25,  # QUARTER
        0.1,  # DIME
        0.05,  # NICKEL
        0.01,  # PENNY
    ]

    money_count = 0
    for currency in currencies:
        if change == 0:
            return int(money_count)

        if change >= currency:
            change, fractions = calc_change_and_fractions(change, currency)
            money_count += fractions

    return int(money_count)


if __name__ == "__main__":
    main()
